### PR TITLE
feat: Auth Middleware

### LINKED ISSUE #6: Implement Auth as Middleware
Add Auth as a middleware, this way we avoid repeating the same lines for each route.

**Accepttest**
* Implement JWT validation as a middleware
* Consider if it would make sense to be able to validate JWTs from Astro too by changing JWT to being verifiable with a public key (this has to happen in API)

### PR DESCRIPTION
**Pull request by kristianbinau**
## Description

Added Auth Middleware

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: src/middleware.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		locals.userId = null;
+		locals.isAuthenticated = false;
+		return next();
+	}
+
+	// If the token is valid, the user is authenticated.
+	locals.authHeader = authHeader ?? undefined;
+	locals.userId = payload.sub ?? null;
+	locals.isAuthenticated = payload.sub !== undefined;
+	return next();
```
Comment (frederikpyt) [inline src/middleware.ts]: Can be moved inside try catch
```suggestion
	try {
		const { payload } = await jwtVerify(rawToken, publicKey, {
			algorithms: ['EdDSA'],
			clockTolerance: 5,
			requiredClaims: ['sub'],
		});
              locals.authHeader = authHeader ?? undefined;
             	locals.userId = payload.sub ?? null;
             	locals.isAuthenticated = payload.sub !== undefined;
              return next();
	} catch (e) {
		// If the token is invalid, the user is not authenticated.
		locals.userId = null;
		locals.isAuthenticated = false;
		return next();
	}
```

File: src/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { sequence, defineMiddleware } from 'astro:middleware';
```
Comment (frederikpyt) [inline src/middleware.ts]: Filename should indicate what this middleware does

File: src/pages/index.astro [hunk 3]
```diff
@@ -1,24 +1,9 @@
 ---
```
Comment (frederikpyt) [inline src/pages/index.astro:1]: Way cleaner 😎

File: src/pages/u/settings.astro [hunk 4]
```diff
@@ -2,12 +2,13 @@
 import LayoutSidebar from '@layouts/LayoutSidebar.astro';
 import UserSettings from '@components/UserSettings.astro';
 import { url, api } from '@lib/helpers';
+import type { User } from '@lib/types';
```
Comment (frederikpyt) [inline src/pages/u/settings.astro:5]: Unused import?

**Review by frederikpyt — CHANGES_REQUESTED** [review]

File: src/middleware.ts (hunk 2 above)
Comment (kristianbinau) [reply src/middleware.ts]: This file is actually the only middleware file allowed if I understood correctly.
The way we make more middleware's is by adding them to the sequence function in the bottom of the page.

File: src/pages/u/settings.astro (hunk 4 above)
Comment (kristianbinau) [reply src/pages/u/settings.astro:5]: It was forgotten in an earlier commit i guess.. See line 24/25

File: src/pages/index.astro (hunk 3 above)
Comment (kristianbinau) [reply src/pages/index.astro:1]: Yep, it's also 10x more performant

File: src/middleware.ts (hunk 2 above)
Comment (kristianbinau) [reply src/middleware.ts]: nvm, you can create a middleware/index.ts and then include files from there.

File: src/middleware.ts (hunk 1 above)
Comment (kristianbinau) [reply src/middleware.ts]: Yeah, that's cleaner.

File: src/middleware/auth.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+		// If the token is valid, the user is authenticated.
+		locals.authHeader = authHeader ?? undefined;
+		locals.userId = payload.sub ?? null;
+		locals.isAuthenticated = payload.sub !== undefined;
+	} catch (e) {
+		// If the token is invalid, the user is not authenticated.
+		locals.userId = null;
+		locals.isAuthenticated = false;
+		return next();
```
Comment (frederikpyt) [inline src/middleware/auth.ts]: Can be omitted

**Review by frederikpyt — CHANGES_REQUESTED** [review]

File: src/middleware/auth.ts (hunk 5 above)
Comment (kristianbinau) [reply src/middleware/auth.ts]:
```suggestion
```

**Review by frederikpyt — APPROVED** [review]
:rocket:

**Pull request merged**